api/src/main/java/io/druid/data/input/impl/PrefetcheableTextFilesFirehoseFactory.java

```diff
@@ -0,0 +1,372 @@
+  {
+    return cacheFiles;
+  }
+
+  /**
+   * Cache objects in a local disk up to {@link #maxCacheCapacityBytes}.
+   */
+  private void cache()
+  {
+    double totalFetchedBytes = 0;
+    final List<ObjectType> objects = getObjects();
+    try {
+      for (int i = 0; i < objects.size() && totalFetchedBytes < maxCacheCapacityBytes; i++) {
+        final ObjectType object = objects.get(i);
+        LOG.info("Caching object[%s]", object);
+        final File outFile = File.createTempFile("cache-", null, baseDir);
+        totalFetchedBytes += download(object, outFile, 0);
+        cacheFiles.add(new FetchedFile(outFile, isGzipped(object)));
+        nextFetchIndex++;
+      }
+    }
+    catch (Exception e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  /**
+   * Fetch objects to a local disk up to {@link #maxFetchCapacityBytes}.  This method is not thread safe and must be
+   * called by a single thread.  Note that even {@link #maxFetchCapacityBytes} is 0, at least 1 file is always fetched.
+   * This is for simplifying design, and should be improved when our client implementations for cloud storages like S3
+   * support range scan.
+   */
+  private void fetch() throws Exception
+  {
+    final List<ObjectType> objects = getObjects();
+    for (int i = nextFetchIndex; i < objects.size() && fetchedBytes.get() <= maxFetchCapacityBytes; i++) {
+      final ObjectType object = objects.get(i);
+      LOG.info("Fetching object[%s]", object);
+      final File outFile = File.createTempFile("fetch-", null, baseDir);
+      fetchedBytes.addAndGet(download(object, outFile, 0));
+      fetchFiles.put(new FetchedFile(outFile, isGzipped(object)));
+      nextFetchIndex++;
+    }
+  }
+
+  /**
+   * Downloads an object. It retires downloading {@link #maxFetchRetry} times and throws that exception.
+   *
+   * @param object  an object to be downloaded
+   * @param outFile a file which the object data is stored
+   * @param retry   current retry count
+   *
+   * @return number of downloaded bytes
+   *
+   * @throws IOException
+   */
+  private long download(ObjectType object, File outFile, int retry) throws IOException
+  {
+    try (final InputStream is = openStream(object);
+         final CountingOutputStream cos = new CountingOutputStream(new FileOutputStream(outFile))) {
+      IOUtils.copy(is, cos);
+      return cos.getCount();
+    }
+    catch (IOException e) {
+      if (retry < maxFetchRetry) {
+        LOG.error(e, "Failed to download object[%s], retrying (%d of %d)", object, retry + 1, maxFetchRetry);
+        outFile.delete();
+        return download(object, outFile, retry + 1);
+      } else {
+        LOG.error(e, "Failed to download object[%s], retries exhausted, aborting", object);
+        throw e;
+      }
+    }
+  }
+
+  @Override
+  public Firehose connect(StringInputRowParser firehoseParser) throws IOException
+  {
+    if (baseDir == null) {
+      baseDir = Files.createTempDir();
+      baseDir.deleteOnExit();
```

Comment:
Expanding use of `deleteOnExit` here is a little undesirable (see #3923 where we removed others). We did use it in ReplayableFirehoseFactory but that was only used by the IndexTask. Now these firehoses are used by potentially more code.

I wonder if instead, we should either add a `File temporaryDirectory` parameter to `connect` or do a hack similar to #4069, to allow the caller to pass in an appropriate tmp directory. Something nicer than #4069 is probably better, since that approach doesn't work if the firehose is wrapped in any way.